Lib/test/test_logging.py

```diff
@@ -4310,6 +4312,31 @@ def test_issue_89047(self):
             s = f.format(r)
             self.assertNotIn('.1000', s)
 
+    def test_issue_102402_msecs_has_no_floating_point_precision_loss(self):
+        tests = (
+            # time_ns is approx. 2023-03-04 04:25:20 UTC
+            # (time_ns, expected_msecs_value)
+            (1_677_902_297_100_000_000, 100.0),  # exactly 100ms
+            (1_677_903_920_999_998_503, 999.0),  # check truncating doesn't round
+            (1_677_903_920_000_998_503, 0.0),  # check truncating doesn't round
+        )
+        for ns, want in tests:
+            with patch('time.time_ns') as patched_ns:
+                patched_ns.return_value = ns
+                record = logging.makeLogRecord({'msg': 'test'})
+            self.assertEqual(record.msecs, want)
```

Comment:
Added. I went with `self.assertEqual(record.created, ns / 1e9)` instead of adjusting `tests` to be a 3-tuple of `(ns: int, want_msecs: float, want_created: float)`. LMK if you'd prefer the 3-tuple

Reply:
It is all what I needed. Thank you.